source/extensions/filters/listener/proxy_protocol/proxy_protocol.h

```diff
@@ -23,19 +23,54 @@ namespace ProxyProtocol {
 using KeyValuePair =
     envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol::KeyValuePair;
 
+enum ProxyProtocolVersion { Unknown = 0, V1 = 1, V2 = 2 };
+
+enum class ReadOrParseState { Done, TryAgainLater, Error, SkipFilter, Denied };
+
 /**
- * All stats for the proxy protocol. @see stats_macros.h
+ * Non-versioned general stats for the filter.
+ * Kept for backwards compatibility.
+ * @see stats_macros.h
  */
 // clang-format off
-#define ALL_PROXY_PROTOCOL_STATS(COUNTER)                                                          \
+#define GENERAL_PROXY_PROTOCOL_STATS(COUNTER)                                                          \
   COUNTER(downstream_cx_proxy_proto_error)
 // clang-format on
 
+struct GeneralProxyProtocolStats {
+  GENERAL_PROXY_PROTOCOL_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+/**
+ * Stats reported for each version of the proxy protocol.
+ * @see stats_macros.h
+ */
+// clang-format off
+#define VERSIONED_PROXY_PROTOCOL_STATS(COUNTER)  \
+  COUNTER(allowed)                               \
+  COUNTER(denied)                                \
+  COUNTER(error)
```

Comment:
I don't think it makes sense to have an error counter for the `unknown` version

Reply:
While I generally agree, unfortunately I'm trying to keep this aligned with the pre-existing behavior and legacy stat `downstream_cx_proxy_proto_error`.

With the current implementation, `unknown.error` is incremented whenever `downstream_cx_proxy_proto_error` is incremented. Pre-existing code increments `downstream_cx_proxy_proto_error` when a request comes in with a non-PP signature or a **partial** signature match. For example, `PROXY abc ` matches the `PROXY` portion of PPv1, but does not end in an `\r\n`. The pre-existing implementation treats this as an `unknown.error` (not as a `v1.error`) because it looks for both `PROXY` and `\r\n` before considering it a PPv1.

I agree `unknown.denied` would be a better stat to increment in these cases, but I'm worried it will cause confusion when `downstream_cx_proxy_proto_error` is also incremented. 

One nice thing is that `*. denied` will only be incremented when the new `allowed_versions` is configured. This way, users who don't configure `allowed_versions` don't need to consider any of the `*. denied` stats and can just look at `*.allowed` and `*.error`.

Reply:
I didn't explicitly mention in the above reply: I don't want to change the behavior of `downstream_cx_proxy_proto_error`

Reply:
I agree we can't change the behavior of `downstream_cx_proxy_proto_error`. But given it already exists, why even add `unknown.error`? There's already a stat for it. And I'd argue that `unknown.allowed` and `unknown.denied` are kind of odd; I'm guessing `unknown.allowed` is for any case where `allow no proxy protocol` is enabled and it isn't detected on a connection. What is `unknown.denied`? 

I'm not sure it makes sense to treat `unknown` as a protocol version; it's the the absence of a detected protocol, which isn't the same concept.

Reply:
Let's take a step back, as a Envoy user I want to answer the following questions:

1. "How many requests have PPv1, PPv2, or no PP attached?"
2. "What decision did the filter make for each type of request?"

> I agree we can't change the behavior of downstream_cx_proxy_proto_error. But given it already exists, why even add unknown.error? There's already a stat for it. 

`downstream_cx_proxy_proto_error` tracks **all** errors for v1, v2, and no PP attached. Using this stat alone, there is no way to answer questions like:

1. "How many requests using PPv1 had parse errors?"
2. "How many requests did the filter receive with no PP (assuming `allow_requests_without_proxy_protocol=false`)

For 1, `v1.error` answers it. For 2, `unknown.error` answers it.

> I'm guessing unknown.allowed is for any case where allow no proxy protocol is enabled and it isn't detected on a connection.

Correct.

> What is unknown.denied

It is currently not populated. As mentioned before, I think it's OK to have a stat that isn't populated, as it will never be pushed to the stats sink. It is also documented correctly. LMK if we need to change this.

> I'm not sure it makes sense to treat unknown as a protocol version; it's the the absence of a detected protocol, which isn't the same concept.

Sorry, I'm still not convinced. From the end user's perspective, `unknown` (renamed to `not_found` now) is a valid "version" that we want to keep track of stats for. `unknown.error` and `unknown.allowed` will let the user correctly answer "How many requests did the filter receive with no PP, and what decision did the filter make"?

Maybe you want me to create a different set of stats for the `not_found` cases that are different from the v1/v2 stats? TBH I feel that increases complexity for the end user to understand, but I am open to it.

Reply:
> It is currently not populated. As mentioned before, I think it's OK to have a stat that isn't populated, as it will never be pushed to the stats sink. It is also documented correctly. LMK if we need to change this.

That's not really correct. There are many stats formats supported by envoy. For example, prometheus will scrape all of the stats; whether it's ever used or not isn't considered. So there is a cost to it. Also, it's confusing; there's a documented stat, that can never be incremented.

There is a cost (in terms of TSDB space, and memory/cpu overhead) of every stat. So producing some that are not useful is undesirable.

From that point, the other thing that feels undesirable here is that stats are produced for things that can't happen, in the current configuration. Depending on the value of `allow_requests_without_proxy_protocol`, some stats can/cannot get populated. Overall, I think that `allow_requests_without_proxy_protocol` should be considered a fringe case; it's explicitly disallowed by the spec. It mostly shouldn't be used. So producing stats for this use case, even when it's not configured, seems undesirable.

> Maybe you want me to create a different set of stats for the not_found cases that are different from the v1/v2 stats? TBH I feel that increases complexity for the end user to understand, but I am open to it.

I'm not convinced yet either way which makes more sense. I appreciate the debate and discussion.

Reply:
> So there is a cost to it. Also, it's confusing; there's a documented stat, that can never be incremented.

SG. I can change the stats structs here to **not** capture these cases.

> Overall, I think that allow_requests_without_proxy_protocol should be considered a fringe case; it's explicitly disallowed by the spec. It mostly shouldn't be used. So producing stats for this use case, even when it's not configured, seems undesirable.

This one I disagree with. We have a use case for it - we are live-migrating traffic to switch from no PP to PPv2. During the short migration period, we will need to enable `allow_requests_without_proxy_protocol`: there is no way to atomically switch Envoy from not parsing PP to parsing PP via filter config otherwise. I imagine other organizations also face the same issue.

When `allow_requests_without_proxy_protocol`, we need to know how many requests without PP are being accepted/denied. I'd rather err on the side of extra stats compared to no stats. IMO

> I'm not convinced yet either way which makes more sense. I appreciate the debate and discussion.

How about we take your suggestion of removing `unknown`/`not_found` from the versioned stats, but we still create separate stats to track them.

```
listener.<stat_prefix>.downstream_cx_proxy_proto.v1.allowed
listener.<stat_prefix>.downstream_cx_proxy_proto.v1.denied
listener.<stat_prefix>.downstream_cx_proxy_proto.v1.error
listener.<stat_prefix>.downstream_cx_proxy_proto.v2.allowed
listener.<stat_prefix>.downstream_cx_proxy_proto.v2.denied
listener.<stat_prefix>.downstream_cx_proxy_proto.v2.error

# Incremented when `allow_requests_without_proxy_protocol=true` and request doesn't have PP.
listener.<stat_prefix>.downstream_cx_proxy_proto.not_found.allowed
# Incremented when `allow_requests_without_proxy_protocol=false` and request doesn't have PP.
listener.<stat_prefix>.downstream_cx_proxy_proto.not_found.denied
```

WDYT? I am open to renaming the last 2. But TBH I think we need to track them. I'd rather err on the site of tracking extra stats vs not tracking enough, and I believe this is one of Envoy's principles.

Reply:
Alright, I think this sounds good overall.

I don't really like allowed/disallowed as the stat names, because it sounds like we're making an rbac/auth decision or something, when in reality we're just checking if they speak the protocol we expect. 

Looking for prior art in other listener filters, both the tls_inspector and http_inspector use `_found`. What if we had

```
....v1.found
....v1.disallowed
....v1.error
````
and in the case where it is disallowed, both found and disallowed are incremented.

Reply:
That's funny, I was referencing the RBAC filter when creating the stats here :)

Thanks for sharing the prior art, agreed those new names sound much better. Will update

Reply:
Ok, here are the new stat names. `found_found` has been moved out from the versioned stats into its own category.

```
listener.stat_prefix.downstream_cx_proxy_proto.not_found_disallowed
listener.stat_prefix.downstream_cx_proxy_proto.not_found_allowed
listener.stat_prefix.downstream_cx_proxy_proto.v1.found
listener.stat_prefix.downstream_cx_proxy_proto.v1.disallowed
listener.stat_prefix.downstream_cx_proxy_proto.v1.error
listener.stat_prefix.downstream_cx_proxy_proto.v2.found
listener.stat_prefix.downstream_cx_proxy_proto.v2.disallowed
listener.stat_prefix.downstream_cx_proxy_proto.v2.error
```

Let me know if you have concerns on `not_found_disallowed` and `not_found_allowed`.

Reply:
PTAL @ggreenway